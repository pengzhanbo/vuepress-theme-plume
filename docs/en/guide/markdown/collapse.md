---
title: Collapsible panel
icon: carbon:collapse-categories
createTime: 2025/03/25 10:13:04
permalink: /en/guide/markdown/collapse/
badge:
  type: tip
  text: 1.0.0-rc.137 +
---

## Overview

In Markdown, the `::: collapse` container, combined with Markdown unordered list syntax, can be used to create ==collapsible panels==.

- It supports setting the mode to **accordion** using the `accordion` option.

## Enable

This feature is not enabled by default. You need to enable it in the `theme` configuration.

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      collapse: true, // [!code ++]
    }
  })
})
```

## Usage

In Markdown, use the `::: collapse` container with Markdown unordered list syntax, where each item represents a separate collapsible area.

```md title="collapse.md"
::: collapse
- Title 1        <!-- Title, click to control expand/collapse -->
                <!-- There must be an empty line between the title and the content -->
  Content          <!-- Content, the collapsible area -->

- Title 2

  Content
:::
```

For each item in the list:

- From the **first line** to the **first empty line** is the **title**.
- **After the first empty line**: The main content.

:::important Please note the correct indentation
:::

**A simple example:**

**Input:**

```md
::: collapse
- Title 1

  Main content

- Title 2

  Main content
:::
```

**Output:**

::: collapse

- Title 1

  Main content

- Title 2

  Main content
:::

## Configuration

After the `::: collapse` container syntax, you can add configuration options:

- `accordion`: Sets the collapsible panels to ==accordion== mode. In accordion mode, only one panel can be expanded at a time. Clicking another panel will close the previously opened one.
- `expand`: Expands all panels by default. This option is invalid in accordion mode.

Before the title of each list item, you can use special markers `:+` or `:-` to set the initial state of the item to **expanded** or **collapsed**.

## Examples

### Basic Usage

**Input:**

```md
::: collapse
- Title 1

  Main content

- Title 2

  Main content
:::
```

**Output:**

::: collapse

- Title 1

  Main content

- Title 2

  Main content
  :::

### Expand All by Default

Add the `expand` option to expand all panels by default.

**Input:**

```md /expand/
::: collapse expand
- Title 1

  Main content

- Title 2

  Main content
:::
```

**Output:**

::: collapse expand

- Title 1

  Main content

- Title 2

  Main content
  :::

### Accordion Mode

Add the `accordion` option to set the mode to accordion, allowing only one panel to be expanded at a time. Clicking another panel will close the previously opened one.

```md /accordion/
::: collapse accordion
- Title 1

  Main content

- Title 2

  Main content

- Title 3

  Main content
:::
```

**Output:**

::: collapse accordion

- Title 1

  Main content

- Title 2

  Main content

- Title 3

  Main content
  :::

### Expand Items with `:+`

By default, all collapsible panels are closed. You can use the `:+` marker to set the initial state of an item to expanded.

**Input:**

```md /:+/
::: collapse
- Title 1

  Main content

- :+ Title 2

  Main content

- :+ Title 3

  Main content
:::
```

**Output:**

::: collapse

- Title 1

  Main content

- :+ Title 2

  Main content

- :+ Title 3

  Main content
  :::

### Collapse Items with `:-`

When the `expand` option is configured, all panels are expanded by default. You can use the `:-` marker to set the initial state of an item to collapsed.

**Input:**

```md /:-/
::: collapse expand
- Title 1

  Main content

- :- Title 2

  Main content

- Title 3

  Main content
:::
```

**Output:**

::: collapse expand

- Title 1

  Main content

- :- Title 2

  Main content

- Title 3

  Main content
  :::
