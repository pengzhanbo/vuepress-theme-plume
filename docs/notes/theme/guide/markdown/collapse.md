---
title: 折叠面板
icon: carbon:collapse-categories
createTime: 2025/03/22 22:27:22
permalink: /guide/markdown/collapse/
badge:
  type: tip
  text: 1.0.0-rc.137 +
---

## 概述

在 markdown 中，使用 `::: collapse` 容器，包含 markdown 无序列表语法，实现 ==折叠面板== 。

- 支持通过 `accordion` 设置为 ==手风琴== 模式

## 启用

该功能默认不启用，你需要在 `theme` 配置中启用。

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      collapse: true, // [!code ++]
    }
  })
})
```

## 使用

在 markdown 中，使用 `::: collapse` 容器，包含 markdown 无序列表语法，每一项为一个单独的可折叠区域。

```md title="collapse.md"
::: collapse
- 标题 1        <!-- 标题，点击控制 展开/折叠 -->
                <!-- 标题与内容必须空一行 -->
  内容          <!-- 内容，被折叠的区域-->

- 标题 2

  内容
:::
```

对于列表的每一个项：

- 从 __首行开始__ 到 __首个空行__，均为 __标题__

- __首个空行之后__: 正文内容

:::important 请注意添加正确的缩进
:::

__一个简单的例子：__

__输入：__

```md
::: collapse
- 标题 1

  正文内容

- 标题 2

  正文内容
:::
```

__输出：__

::: collapse

- 标题 1

  正文内容

- 标题 2

  正文内容
:::

## 配置

在 `::: collapse` 容器语法之后，跟随配置项：

- `accordion` ：折叠面板设置为 ==手风琴== 模式，在手风琴模式下，只允许展开一个面板，点击其他面板会关闭之前的面板。
- `expand` ：默认展开面板，在手风琴模式下无效。

在列表项，标题之前，可通过特殊标记 `:+` / `:-` 来设置当前项是否 __展开 / 折叠__。

## 示例

### 基本用法

__输入：__

```md
::: collapse
- 标题 1

  正文内容

- 标题 2

  正文内容
:::
```

__输出：__

::: collapse

- 标题 1

  正文内容

- 标题 2

  正文内容
:::

### 默认全部展开

添加 `expand` 选项，默认展开所有面板

__输入：__

```md /expand/
::: collapse expand
- 标题 1

  正文内容

- 标题 2

  正文内容
:::
```

__输出：__

::: collapse expand

- 标题 1

  正文内容

- 标题 2

  正文内容
:::

### 手风琴模式

添加 `accordion` 选项，设置为手风琴模式，只允许展开一个面板，点击其他面板会关闭之前的面板

```md /accordion/
::: collapse accordion
- 标题 1

  正文内容

- 标题 2

  正文内容

- 标题 3

  正文内容
:::
```

__输出：__

::: collapse accordion

- 标题 1

  正文内容

- 标题 2

  正文内容

- 标题 3

  正文内容
:::

### `:+` 标记项为展开

折叠面板默认全部关闭，可以使用 `:+` 标记项初始状态为展开。

__输入：__

```md /:+/
::: collapse
- 标题 1

  正文内容

- :+ 标题 2

  正文内容

- :+ 标题 3

  正文内容
:::
```

__输出：__

::: collapse

- 标题 1

  正文内容

- :+ 标题 2

  正文内容

- :+ 标题 3

  正文内容
:::

### `:-` 标记项为折叠

折叠面板配置 `expand` 时默认全部展开，可以使用 `:-` 标记项初始状态为折叠。

__输入：__

```md /:-/
::: collapse expand
- 标题 1

  正文内容

- :- 标题 2

  正文内容

- 标题 3

  正文内容
:::
```

__输出：__

::: collapse expand

- 标题 1

  正文内容

- :- 标题 2

  正文内容

- 标题 3

  正文内容
:::
