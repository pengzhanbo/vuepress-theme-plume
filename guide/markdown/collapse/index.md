---
url: /guide/markdown/collapse/index.md
---
## 概述

在 markdown 中，使用 `::: collapse` 容器，包含 markdown 无序列表语法，实现 ==折叠面板== 。

* 支持通过 `accordion` 设置为 ==手风琴== 模式

## 启用

该功能默认不启用，你需要在 `theme` 配置中启用。

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdown: {
      collapse: true, // [!code ++]
    }
  })
})
```

## 使用

在 markdown 中，使用 `::: collapse` 容器，包含 markdown 无序列表语法，每一项为一个单独的可折叠区域。

```md title="collapse.md"
::: collapse
- 标题 1        <!-- 标题，点击控制 展开/折叠 -->
                <!-- 标题与内容必须空一行 -->
  内容          <!-- 内容，被折叠的区域-->

- 标题 2

  内容
:::
```

对于列表的每一个项：

* 从 **首行开始** 到 **首个空行**，均为 **标题**

* **首个空行之后**: 正文内容

:::important 请注意添加正确的缩进
:::

**一个简单的例子：**

**输入：**

```md
::: collapse
- 标题 1

  正文内容

- 标题 2

  正文内容
:::
```

**输出：**

::: collapse

* 标题 1

  正文内容

* 标题 2

  正文内容
  :::

## 配置

在 `::: collapse` 容器语法之后，跟随配置项：

* `accordion` ：折叠面板设置为 ==手风琴== 模式，在手风琴模式下，只允许展开一个面板，点击其他面板会关闭之前的面板。
* `expand` ：默认展开面板，在手风琴模式下无效。

在列表项，标题之前，可通过特殊标记 `:+` / `:-` 来设置当前项是否 **展开 / 折叠**。

## 示例

### 基本用法

**输入：**

```md
::: collapse
- 标题 1

  正文内容

- 标题 2

  正文内容
:::
```

**输出：**

::: collapse

* 标题 1

  正文内容

* 标题 2

  正文内容
  :::

### 默认全部展开

添加 `expand` 选项，默认展开所有面板

**输入：**

```md /expand/
::: collapse expand
- 标题 1

  正文内容

- 标题 2

  正文内容
:::
```

**输出：**

::: collapse expand

* 标题 1

  正文内容

* 标题 2

  正文内容
  :::

### 手风琴模式

添加 `accordion` 选项，设置为手风琴模式，只允许展开一个面板，点击其他面板会关闭之前的面板

```md /accordion/
::: collapse accordion
- 标题 1

  正文内容

- 标题 2

  正文内容

- 标题 3

  正文内容
:::
```

**输出：**

::: collapse accordion

* 标题 1

  正文内容

* 标题 2

  正文内容

* 标题 3

  正文内容
  :::

### `:+` 标记项为展开

折叠面板默认全部关闭，可以使用 `:+` 标记项初始状态为展开。

**输入：**

```md /:+/
::: collapse
- 标题 1

  正文内容

- :+ 标题 2

  正文内容

- :+ 标题 3

  正文内容
:::
```

**输出：**

::: collapse

* 标题 1

  正文内容

* :+ 标题 2

  正文内容

* :+ 标题 3

  正文内容
  :::

### `:-` 标记项为折叠

折叠面板配置 `expand` 时默认全部展开，可以使用 `:-` 标记项初始状态为折叠。

**输入：**

```md /:-/
::: collapse expand
- 标题 1

  正文内容

- :- 标题 2

  正文内容

- 标题 3

  正文内容
:::
```

**输出：**

::: collapse expand

* 标题 1

  正文内容

* :- 标题 2

  正文内容

* 标题 3

  正文内容
  :::
