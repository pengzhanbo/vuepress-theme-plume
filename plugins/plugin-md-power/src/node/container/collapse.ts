/**
 * ::: collapse accordion
 * - + 标题
 *   内容
 * - - 标题
 *   内容
 * :::
 */
import type Token from 'markdown-it/lib/token.mjs'
import type { Markdown } from 'vuepress/markdown'
import { resolveAttrs } from '../utils/resolveAttrs.js'
import { stringifyAttrs } from '../utils/stringifyAttrs.js'
import { createContainerPlugin } from './createContainer.js'

interface CollapseMeta {
  accordion?: boolean
  expand?: boolean
}

interface CollapseItemMeta {
  expand?: boolean
  index?: number
}

export function collapsePlugin(md: Markdown): void {
  createContainerPlugin(md, 'collapse', {
    before: (info, tokens, index) => {
      const { attrs } = resolveAttrs<CollapseMeta>(info)
      const idx = parseCollapse(tokens, index, attrs)
      const { accordion } = attrs

      return `<VPCollapse${stringifyAttrs({ accordion, index: idx })}>`
    },
    after: () => `</VPCollapse>`,
  })
  md.renderer.rules.collapse_item_open = (tokens, idx) => {
    const token = tokens[idx]
    const { expand, index } = token.meta as CollapseItemMeta
    return `<VPCollapseItem${stringifyAttrs({ expand, index })}>`
  }
  md.renderer.rules.collapse_item_close = () => '</VPCollapseItem>'
  md.renderer.rules.collapse_item_title_open = () => '<template #title>'
  md.renderer.rules.collapse_item_title_close = () => '</template>'
}

function parseCollapse(tokens: Token[], index: number, attrs: CollapseMeta): number | void {
  const listStack: number[] = [] // 记录列表嵌套深度
  let idx = -1 // 记录当前列表项下标
  let defaultIndex: number | undefined
  let hashExpand = false
  for (let i = index + 1; i < tokens.length; i++) {
    const token = tokens[i]
    if (token.type === 'container_collapse_close') {
      break
    }
    // 列表层级追踪
    if (token.type === 'bullet_list_open') {
      listStack.push(0) // 每个新列表初始层级为0
      if (listStack.length === 1)
        token.hidden = true
    }
    else if (token.type === 'bullet_list_close') {
      listStack.pop()
      if (listStack.length === 0)
        token.hidden = true
    }
    else if (token.type === 'list_item_open') {
      const currentLevel = listStack.length
      // 仅处理根级列表项（层级1）
      if (currentLevel === 1) {
        token.type = 'collapse_item_open'
        tokens[i + 1].type = 'collapse_item_title_open'
        tokens[i + 3].type = 'collapse_item_title_close'

        idx++

        const inlineToken = tokens[i + 2]
        const firstToken = inlineToken.children![0]
        let flag: string = ''
        let expand: boolean | undefined
        if (firstToken.type === 'text') {
          firstToken.content = firstToken.content.trim().replace(/^:[+\-]\s*/, (match) => {
            flag = match.trim()
            return ''
          })
        }
        if (attrs.accordion) {
          if (!hashExpand && flag === ':+') {
            expand = hashExpand = true
            defaultIndex = idx
          }
        }
        else if (flag === ':+') {
          expand = true
        }
        else if (flag === ':-') {
          expand = false
        }
        else {
          expand = !!attrs.expand
        }

        token.meta = {
          index: idx,
          expand,
        } as CollapseItemMeta
      }
    }
    else if (token.type === 'list_item_close') {
      const currentLevel = listStack.length
      if (currentLevel === 1) {
        token.type = 'collapse_item_close'
      }
    }
  }
  if (attrs.accordion && attrs.expand && !hashExpand) {
    defaultIndex = 0
  }
  return defaultIndex
}
