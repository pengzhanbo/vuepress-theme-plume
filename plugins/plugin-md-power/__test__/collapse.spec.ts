import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { collapsePlugin } from '../src/node/container/collapse.js'

describe('collapsePlugin', () => {
  const md = new MarkdownIt().use(collapsePlugin)
  it('should work', () => {
    const code = `\
::: collapse
- :+ 标题

  内容

- :- \`code\`标题

  内容
  - 列表 1
  - 列表 2

- \`code\` 标题

  内容
:::
`
    expect(md.render(code)).toMatchSnapshot()
  })

  it('should work with expand', () => {
    const code = `\
::: collapse expand
- 标题

  内容

- 标题

  内容

- :- 标题

  内容
:::
`
    expect(md.render(code)).toMatchSnapshot()
  })

  it('should work with accordion', () => {
    const code = `\
::: collapse accordion
- 标题

  内容

- 标题

  内容

- 标题

  内容
:::

::: collapse accordion expand
- 标题

  内容

- 标题

  内容

- 标题

  内容
:::

::: collapse accordion
- 标题

  内容

- :+ 标题

  内容

- 标题

  内容
:::
`
    expect(md.render(code)).toMatchSnapshot()
  })
})
