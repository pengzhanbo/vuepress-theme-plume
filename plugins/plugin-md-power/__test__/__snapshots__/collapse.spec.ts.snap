// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`collapsePlugin > should work 1`] = `
"<VPCollapse><VPCollapseItem expand :index="0"><template #title>标题</template><p>内容</p>
</VPCollapseItem><VPCollapseItem :index="1"><template #title><code>code</code>标题</template><p>内容</p>
<ul>
<li>列表 1</li>
<li>列表 2</li>
</ul>
</VPCollapseItem><VPCollapseItem :index="2"><template #title><code>code</code> 标题</template><p>内容</p>
</VPCollapseItem></VPCollapse>"
`;

exports[`collapsePlugin > should work with accordion 1`] = `
"<VPCollapse accordion><VPCollapseItem :index="0"><template #title>标题</template><p>内容</p>
</VPCollapseItem><VPCollapseItem :index="1"><template #title>标题</template><p>内容</p>
</VPCollapseItem><VPCollapseItem :index="2"><template #title>标题</template><p>内容</p>
</VPCollapseItem></VPCollapse><VPCollapse accordion :index="0"><VPCollapseItem :index="0"><template #title>标题</template><p>内容</p>
</VPCollapseItem><VPCollapseItem :index="1"><template #title>标题</template><p>内容</p>
</VPCollapseItem><VPCollapseItem :index="2"><template #title>标题</template><p>内容</p>
</VPCollapseItem></VPCollapse><VPCollapse accordion :index="1"><VPCollapseItem :index="0"><template #title>标题</template><p>内容</p>
</VPCollapseItem><VPCollapseItem expand :index="1"><template #title>标题</template><p>内容</p>
</VPCollapseItem><VPCollapseItem :index="2"><template #title>标题</template><p>内容</p>
</VPCollapseItem></VPCollapse>"
`;

exports[`collapsePlugin > should work with expand 1`] = `
"<VPCollapse><VPCollapseItem expand :index="0"><template #title>标题</template><p>内容</p>
</VPCollapseItem><VPCollapseItem expand :index="1"><template #title>标题</template><p>内容</p>
</VPCollapseItem><VPCollapseItem :index="2"><template #title>标题</template><p>内容</p>
</VPCollapseItem></VPCollapse>"
`;
